### PR TITLE
Use diagnostic or language items instead of paths

### PR DESCRIPTION
**Pull request by Jarcho**
I think that gets everything except ones used in a list of paths to check.

changelog: none

### CODE REVIEW

File: clippy_lints/src/lifetimes.rs [hunk 1]
```diff
@@ -361,7 +360,7 @@ impl<'a, 'tcx> Visitor<'tcx> for RefVisitor<'a, 'tcx> {
         let trait_ref = &poly_tref.trait_ref;
         if CLOSURE_TRAIT_BOUNDS
             .iter()
-            .any(|trait_path| trait_ref.trait_def_id() == get_trait_def_id(self.cx, trait_path))
+            .any(|&item| trait_ref.trait_def_id() == self.cx.tcx.lang_items().items()[item as usize])
```
Comment (camsteffen) [inline clippy_lints/src/lifetimes.rs:363]:
```suggestion
            .any(|&item| Some(trait_ref.trait_def_id()) == self.cx.tcx.lang_items().require(item).ok()
```

File: clippy_lints/src/types.rs [hunk 2]
```diff
@@ -287,37 +287,55 @@ impl<'tcx> LateLintPass<'tcx> for Types {
...
-            return Some(ty.span);
-        }
+/// Checks if the first type parameter is a lang item.
+fn is_ty_param_lang_item(cx: &LateContext<'_>, qpath: &QPath<'tcx>, item: LangItem) -> Option<&'tcx hir::Ty<'tcx>> {
+    let ty = get_qpath_generic_tys(qpath).next()?;
+
+    if let TyKind::Path(qpath) = &ty.kind {
+        cx.qpath_res(qpath, ty.hir_id)
+            .opt_def_id()
+            .and_then(|id| (cx.tcx.lang_items().items()[item as usize] == Some(id)).then(|| ty))
```
Comment (camsteffen) [inline clippy_lints/src/types.rs:297]:
```suggestion
            .and_then(|id| (cx.tcx.lang_items().require(item).ok() == Some(id)).then(|| ty))
```

**Review by camsteffen — CHANGES_REQUESTED** [review]

**Comment by phansch** [comment]
Minus the comments from @camsteffen, this looks good to me, thanks!

**Comment by phansch** [comment]
woops, misclicked :sweat_smile:

**Comment by matthiaskrgr** [comment]
Is this expected to have a direct effect on lint results?
I'm asking because the change causes ~160 `expl_impl_clone_on_copy` warnings to appear in the lintcheck logs.

Most of them seem to be caused by macro usage

`target/lintcheck/sources/libc-0.2.81/src/macros.rs:84:9`

````rust
#[allow(unused_macros)]
macro_rules! s {
    ($($(#[$attr:meta])* pub $t:ident $i:ident { $($field:tt)* })*) => ($(
        s!(it: $(#[$attr])* pub $t $i { $($field)* });
    )*);
    (it: $(#[$attr:meta])* pub union $i:ident { $($field:tt)* }) => (
        compile_error!("unions cannot derive extra traits, use s_no_extra_traits instead");
    );
    (it: $(#[$attr:meta])* pub struct $i:ident { $($field:tt)* }) => (
        __item! {
            #[repr(C)]
            #[cfg_attr(feature = "extra_traits", derive(Debug, Eq, Hash, PartialEq))]
            #[allow(deprecated)]
            $(#[$attr])*
            pub struct $i { $($field)* }
        }
        #[allow(deprecated)]
        impl ::Copy for $i {}
        #[allow(deprecated)]
        impl ::Clone for $i { // <- HERE
            fn clone(&self) -> $i { *self }
        }
    );
}
````

````
warning: you are implementing `Clone` explicitly on a `Copy` type
    --> src/macros.rs:84:9
     |
84   | /         impl ::Clone for $i {
85   | |             fn clone(&self) -> $i { *self }
86   | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     |
note: consider deriving `Clone` or removing `Copy`
    --> src/macros.rs:84:9
     |
84   | /         impl ::Clone for $i {
85   | |             fn clone(&self) -> $i { *self }
86   | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
     = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: you are implementing `Clone` explicitly on a `Copy` type
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     |
note: consider deriving `Clone` or removing `Copy`
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
     = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: you are implementing `Clone` explicitly on a `Copy` type
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     |
note: consider deriving `Clone` or removing `Copy`
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
     = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

warning: you are implementing `Clone` explicitly on a `Copy` type
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     |
note: consider deriving `Clone` or removing `Copy`
    --> src/macros.rs:120:9
     |
120  | /         impl ::Clone for $i {
121  | |             fn clone(&self) -> $i { *self }
122  | |         }
     | |_________^
     |
    ::: src/unix/linux_like/linux/mod.rs:3502:1
     |
3502 |   expand_align!();
     |   ---------------- in this macro invocation
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#expl_impl_clone_on_copy
     = note: this warning originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

````

File: clippy_lints/src/derive.rs [hunk 3]
```diff
@@ -293,7 +293,7 @@ fn check_ord_partial_ord<'tcx>(
 
 /// Implementation of the `EXPL_IMPL_CLONE_ON_COPY` lint.
 fn check_copy_clone<'tcx>(cx: &LateContext<'tcx>, item: &Item<'_>, trait_ref: &TraitRef<'_>, ty: Ty<'tcx>) {
-    if match_path(&trait_ref.path, &paths::CLONE_TRAIT) {
+    if cx.tcx.lang_items().clone_trait() == trait_ref.trait_def_id() {
```
Comment (flip1995) [inline clippy_lints/src/derive.rs:296]: Maybe comparing `def_id`s isn't equivalent to the `match_path` method?

Comment (Jarcho) [reply clippy_lints/src/derive.rs:296]: The lang item might not be defined at the point. I didn't think that would actually happen though. I'll have a pr in a bit.

**Comment by flip1995** [comment]
Yeah, I agree. Triggering in the above example makes sense to me. I also think triggering in local macros is fine.

**Pull request merged**